hcl/ast/merged_block.go

```diff
@@ -39,31 +43,57 @@ type MergedBlock struct {
 	RawBlocks map[string]Blocks
 }
 
+// BlockType represents a block type.
+type BlockType string
+
+// LabelBlockType represents a labelled block type.
+type LabelBlockType struct {
+	Type   BlockType // Type of the block
+	Labels string    // Labels are comma separated.
+}
+
 // MergedBlocks maps the block name to the MergedBlock.
 type MergedBlocks map[string]*MergedBlock
 
+// MergedLabelBlocks maps the block labels/types to the MergedBlock.
+type MergedLabelBlocks map[LabelBlockType]*MergedBlock
+
 // NewMergedBlock creates a new MergedBlock of type typ.
-func NewMergedBlock(typ string) *MergedBlock {
+func NewMergedBlock(typ string, labels []string) *MergedBlock {
 	return &MergedBlock{
-		Type:       typ,
+		Type:       BlockType(typ),
+		Labels:     strings.Join(labels, "."),
 		Attributes: make(Attributes),
 		Blocks:     make(map[string]*MergedBlock),
 		RawBlocks:  make(map[string]Blocks),
 	}
 }
 
+// NewLabelBlockType returns a new LabelBlockType.
+func NewLabelBlockType(typ string, labels []string) LabelBlockType {
+	return LabelBlockType{
+		Type:   BlockType(typ),
+		Labels: strings.Join(labels, "."),
+	}
+}
+
 // MergeBlock recursively merges the other block into this one.
-func (mb *MergedBlock) MergeBlock(other *Block) error {
+func (mb *MergedBlock) MergeBlock(other *Block, isLabelled bool) error {
 	errs := errors.L()
-
-	// Currently all merged blocks do not support labels.
-	// This should not be handled here if changed in the future.
-	if len(other.Labels) > 0 {
+	if !isLabelled && len(other.Labels) > 0 {
 		errs.Append(errors.E(other.LabelRanges, "block type %q does not support labels"))
+	} else {
+		otherLabels := strings.Join(other.Labels, ".")
```

Comment:
label building is also here, so maybe the extract could be used here too if it happens.